### DISCUSSION TITLE
How to set up an external host for a provider using pact python?

### DISCUSSION THREAD

**Discussion by alexmonteirocastro** [pact-foundation/pact-python#402] [Q&A] [answered]
Hello, I need some help,

I am trying to build a contract test for 2 backend services

Here's a high level overview of how it works

Main backend interacts with a specific_data service

So my backend is the consumer and the specific_data_service is the provider

When I send a PUT request to the endpoint `/api/sepecific_data/{specific_uuid}/data` then the backend calls the `update_specific_data()` function

The `update_specific_data()` function then sends a PUT request to the external specific_data service which is hosted externally, the base url being something like `https://specific-data-service.run.app/`

My question is, **how do I set up my provider url within the consumer test?**
---

If I set it like

```
pact = Consumer('specific-data-consumer-python').has_pact_with(Provider('specific-data-provider-python'), host_name='http://127.0.0.1/', port=1234, pact_dir="./pacts", log_dir="./logs")
```

Then when I run

```
(pact
         .given('Data exists')
         .upon_receiving('a request for updating data')
         .with_request(method='put', path=f'/api/sepecific_data/{specific_uuid}/data', body=DATA_CHANGED)
         .will_respond_with(200, body=UPDATED_DATA))
    
    with pact:
            result = update_specific_data(args)
            assert result == expected
```

The request isn't actualy being intercepted and tries to call the actual external API

---

Is there a way to set my provider like?

```
pact = Consumer('specific-data-consumer-python').has_pact_with(Provider('specfic-data-provider-python'), host_name='https://specific-data-service.run.app/', pact_dir="./pacts", log_dir="./logs")
```
❓ 

Sorry for the super long text

I really hope someone can help me 🙏

**Comment by mefellows** [comment, marked as answer]
Before I answer, the best way to think of a Pact test is a _unit test_ for your API client. Once you're in this frame of mind, things get a bit easier to understand. Unit tests are easier to reason about - because they are specific, targeted etc. - and give you more control over the test. Pact is a tool that is in contrast to techniques that involve testing live systems, because of the problems (see this series of [videos](https://www.youtube.com/playlist?list=PLwy9Bnco-IpfZ72VQ7hce8GicVZs7nm0i) for more on this).

> The update_specific_data() function then sends a PUT request to the external specific_data service which is hosted externally, the base url being something like https://specific-data-service.run.app/

In Pact tests (because they are unit tests), you need to isolate the provider. You should stub the downstream `specific_data` service in this case (and then create a pact between this provider (which would then be a consumer) and `specific_data`.

> The request isn't actualy being intercepted and tries to call the actual external API

Pact doesn't do this (although I think that would be nice to try). You need to send your requests to the pact mock service, which you have configured to run on port `1234`.

```
result = update_specific_data(args) # <- this function needs to be aware of the correct host to call - i.e. the pact mock service
```
